gloo/transport/uv/pair.cc

```diff
@@ -0,0 +1,530 @@
+// Locking requirements: caller must hold instance mutex.
+//
+void Pair::sendNotifySendReady(uint64_t tag, size_t length) {
+  Op op;
+  op.preamble.nbytes = sizeof(op.preamble);
+  op.preamble.opcode = Op::NOTIFY_SEND_READY;
+  op.preamble.tag = tag;
+  op.preamble.length = length;
+  writeOp(std::move(op));
+}
+
+// Send notification to peer that there is a pending recv operation.
+//
+// Threading: called from either user thread or event loop thread.
+// Locking requirements: caller must hold instance mutex.
+//
+void Pair::sendNotifyRecvReady(uint64_t tag, size_t length) {
+  Op op;
+  op.preamble.nbytes = sizeof(op.preamble);
+  op.preamble.opcode = Op::NOTIFY_RECV_READY;
+  op.preamble.tag = tag;
+  op.preamble.length = length;
+  writeOp(std::move(op));
+}
+
+// Send unbound buffer to peer.
+//
+// Threading: called from either user thread or event loop thread.
+// Locking requirements: caller must hold instance mutex.
+//
+void Pair::sendUnboundBuffer(
+    uint64_t tag,
+    NonOwningPtr<UnboundBuffer> buf,
+    size_t offset,
+    size_t length) {
+  Op op;
+  op.preamble.nbytes = sizeof(op.preamble) + length;
+  op.preamble.opcode = Op::SEND_UNBOUND_BUFFER;
+  op.preamble.tag = tag;
+  op.preamble.length = length;
+  op.buf = std::move(buf);
+  op.offset = offset;
+  op.length = length;
+  writeOp(std::move(op));
+}
+
+// Send from the specified buffer to remote side of pair.
+void Pair::send(
+    transport::UnboundBuffer* tbuf,
+    uint64_t tag,
+    size_t offset,
+    size_t nbytes) {
+  auto buf = static_cast<UnboundBuffer*>(tbuf)->getWeakNonOwningPtr();
+
+  GLOO_ENFORCE_GE(offset, 0);
+  GLOO_ENFORCE_GE(nbytes, 0);
+  if (nbytes > 0) {
+    GLOO_ENFORCE_LE(offset, tbuf->size);
+    GLOO_ENFORCE_LE(nbytes, tbuf->size - offset);
+  }
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // TODO(check if the pair is in the right state)
+
+  // Execute this send if there is a remote pending receive.
+  ContextMutator mutator(*context_, tag, rank_);
+  if (mutator.getRemotePendingRecv() > 0) {
+    // We keep a count of remote pending send and receive operations.
+    // In this code path the remote side hasn't seen a notification
+    // for this send operation yet so we need to take special care
+    // their count is updated regardless.
+    sendNotifySendReady(tag, nbytes);
+    sendUnboundBuffer(tag, NonOwningPtr<UnboundBuffer>(buf), offset, nbytes);
+    mutator.updateRemotePendingRecv(-1);
+    return;
+  }
+
+  // Notify peer of this pending send.
+  localPendingSend_[tag].emplace_back(std::move(buf), offset, nbytes);
+  sendNotifySendReady(tag, nbytes);
+}
+
+// Receive into the specified buffer from the remote side of pair.
+void Pair::recv(
+    transport::UnboundBuffer* tbuf,
+    uint64_t tag,
+    size_t offset,
+    size_t nbytes) {
+  auto buf = static_cast<UnboundBuffer*>(tbuf)->getWeakNonOwningPtr();
+
+  GLOO_ENFORCE_GE(offset, 0);
+  GLOO_ENFORCE_GE(nbytes, 0);
```

Comment:
can `offset` and `nbytes` ever become negative? If not can we directly use unsigned?

Reply:
They can't. You're right. The checks are superfluous. I copied them from the tcp code, where they were used to initially guard against empty send/recv operations. Later we changed this to allow empty send/recv operations and now they're meaningless. I'll submit a separate PR to remove them from the tcp backend as well.

Reply:
This is fixed in #206.